### PR TITLE
attempt at adding disk space monitoring for rkt

### PR DESCRIPTION
**Pull request by sjpotter**
proof of concept as cadvisor doesn't provide functionality for rkt yet.

not sure I'm happy w/ this, but figured it can spark discussion.

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/20887)

### CODE REVIEW

**Comment by sjpotter** [comment]
@yifan-gu

**Comment by yifan-gu** [comment]
cc @jonboulle @dchen1107 @yujuhong

File: pkg/kubelet/disk_manager.go [hunk 1]
```diff
@@ -35,12 +37,12 @@ const mb = 1024 * 1024
 type diskSpaceManager interface {
 	// Checks the available disk space
 	IsRootDiskSpaceAvailable() (bool, error)
-	IsDockerDiskSpaceAvailable() (bool, error)
+	IsContainerDiskSpaceAvailable() (bool, error)
 }
 
 type DiskSpacePolicy struct {
 	// free disk space threshold for filesystem holding docker images.
```
Comment (jonboulle) [inline pkg/kubelet/disk_manager.go]: update

File: pkg/kubelet/disk_manager.go [hunk 2]
```diff
@@ -111,8 +172,8 @@ func (dm *realDiskSpaceManager) isSpaceAvailable(fsType string, threshold int, f
 }
 
 func validatePolicy(policy DiskSpacePolicy) error {
-	if policy.DockerFreeDiskMB < 0 {
-		return fmt.Errorf("free disk space should be non-negative. Invalid value %d for docker disk space threshold.", policy.DockerFreeDiskMB)
+	if policy.ContainersFreeDiskMB < 0 {
+		return fmt.Errorf("free disk space should be non-negative. Invalid value %d for docker disk space threshold.", policy.ContainersFreeDiskMB)
```
Comment (jonboulle) [inline pkg/kubelet/disk_manager.go]: needs update (docker)

File: pkg/kubelet/disk_manager.go [hunk 3]
```diff
@@ -59,6 +61,65 @@ type realDiskSpaceManager struct {
...
+		},
+		rktPath: rktPath,
+	}
+
+	return dm, nil
+}
+
+func getVfsStats(path string) (uint64, uint64, uint64, error) {
+	glog.Info("getVfsStats: enter");
+        var s syscall.Statfs_t
```
Comment (jonboulle) [inline pkg/kubelet/disk_manager.go]: needs a gofmt I guess

**Comment by jonboulle** [comment]
Seems like a reasonable interim solution until the cadvisor stuff is in place.

**Comment by vishh** [comment]
I'd prefer moving this logic into cadvisor. I'd be ok with a short-term solution in cadvisor.

**Comment by sjpotter** [comment]
so there are a few ways to do this

1) make cadvisor able to ditinguih between docker and rkt

i.e. instead of DockerImagesFsInfo() there would be a ContainerrImagesFsInfo and it be smart enough to do its thing based on the container engine

or

2) just add parallel APIs 

i.e. in addition to

DockerImagesFsInfo we would have RktImagesFsInfo

In either approach we could either pass in the rktPath variable or try to figure it out automatically (i think it does someting like that for docker)

**Comment by vishh** [comment]
I prefer option `2` for now.. Once we get around to re-factoring the APIs
to be cross-runtime compatible, we can switch to option `1`. WDYT?

**Comment by sjpotter** [comment]
I dont think its a problem, its essentially what I did already.   Only Q is does it get added to the manager interface in cadvisor itself or just the embded kubelet cadvisor code.

1) RktImagesFsInfo() in kubelet/cadvisor/cadvisor_linux.go actually implements the statfs

or

2) RktImagesFsInfo() uses a call to cadvisor manager code to do the statfs.

**Comment by vishh** [comment]
Option `2`.

**Comment by vishh** [comment]
@sjpotter Is this PR still relevant? If not, can you close this?

**Comment by yifan-gu** [comment]
ping @sjpotter

**Comment by sjpotter** [comment]
not needed anymore

**Pull request closed without merging**